flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PostTransformOperator.java

```diff
@@ -434,23 +436,36 @@ private Optional<DataChangeEvent> processDataChangeEvent(DataChangeEvent dataCha
                         && transformProjectionOptional.get().isValid()) {
                     TransformProjection transformProjection = transformProjectionOptional.get();
                     if (!transformProjectionProcessorMap.containsKey(
-                                    Tuple2.of(tableId, transformProjection))
-                            || !transformProjectionProcessorMap
-                                    .get(Tuple2.of(tableId, transformProjection))
-                                    .hasTableInfo()) {
+                            Tuple2.of(tableId, transformProjection.getProjection()))) {
                         transformProjectionProcessorMap.put(
-                                Tuple2.of(tableId, transformProjection),
+                                Tuple2.of(tableId, transformProjection.getProjection()),
                                 TransformProjectionProcessor.of(
                                         tableInfo,
                                         transformProjection,
                                         timezone,
                                         udfDescriptors,
                                         getUdfFunctionInstances(),
                                         transform.getSupportedMetadataColumns()));
+                    } else if (!transformProjectionProcessorMap
+                            .get(Tuple2.of(tableId, transformProjection.getProjection()))
+                            .hasTableInfo()) {
+                        TransformProjectionProcessor transformProjectionProcessorWithoutTableInfo =
+                                transformProjectionProcessorMap.get(
+                                        Tuple2.of(tableId, transformProjection.getProjection()));
+                        transformProjectionProcessorMap.put(
+                                Tuple2.of(tableId, transformProjection.getProjection()),
+                                TransformProjectionProcessor.of(
```

Comment:
We may have a `copy` method in `TransformProjectionProcessor` to allow us overwriting specific field but keep others intact.

Otherwise, a lot of codes must be modified if we add one more field into it.

Reply:
+1 Use TransformProjectionProcessor#copy() to create a non-shared object seems better.